source/mainnet/smart-contracts/guides/factory-pattern.rst

```diff
@@ -0,0 +1,392 @@
+on the result of the first one. If this two-step process is presented to the end user, it is likely
+to cause confusion. Additionally, two-step process means that the smart contracts have to be robust
+against misuse. For instance, the factory has to deal with the possibility that ``produce`` is
+called with a reference to the wrong type of product contract.
+
+Alternatives to the factory pattern
+===================================
+
+Since there are a number of disadvantages to using the factory pattern on Concordium, it is
+important to consider whether a better alternative is available. One of the following alternatives
+may be suitable, depending on the application.
+
+Construct the product directly
+------------------------------
+
+If the factory does not maintain an on-going relationship with the product,
+then it is generally possible to simply initialize the product entirely in
+the constructor (``init_product``), removing the separate ``initialize`` operation entirely.
+In general, if the factory contract does not need to update its state
+after initializing the product, then it serves no purpose. It would be
+sufficient for the user to construct the product directly in the same way as the factory
+would have.
+
+Register the instance with a contract registry
+----------------------------------------------
+
+A factory contract may serve as a registry of its products.
+The factory pattern enables this, as the factory is responsible for initializing each product
+instance. However, the factory pattern is not in itself necessary for a contract registry.
+Instead, instances could be constructed directly (as discussed above) and subsequently registered
+with a registry contract, which maintains an index of the instances registered with it.
+The registration process could also provide additional initialization information for the contract
+being registered (such as the index it is assigned by the registry).
+
+Use a monolithic contract
+-------------------------
+
+Instead of having separate contract instances for the factory and each product, all of these
+could be combined in a single monolithic contract. In this model, each product is assigned a
+virtual address, and the contract maintains a mapping from virtual addresses to the state of
+each product.
+
+For instance, a CIS2 contract can manage multiple NFTs that each have distinct token IDs.
+Rather than each NFT being its own contract instance (created by a factory), they are simply
+handled as part of the state of the overall CIS2 contract. Here, the token ID actts as a
+virtual address.
+
+The main disadvantage of this approach is that the isolation between the states of each product
+must be enforced by the contract itself. If the state becomes corrupted (due to a bug in the
+contract) then all of the encapsulated products can potentially be affected. With the factory
+pattern, the runtime system of the blockchain enforces isolation. As always, the balance of risks
```

Comment:
This argument here about isolation seems spurious because if all contracts are initialized from the same code, then corruption will likely also lead to corruption in all products.

So this is neither here nor there.